ktor-client/ktor-client-curl/src/io/ktor/client/engine/curl/CurlDriver.kt

```diff
@@ -0,0 +1,47 @@
+package io.ktor.client.engine.curl
+
+import io.ktor.http.*
+import io.ktor.http.content.*
+import kotlinx.cinterop.*
+import kotlin.native.concurrent.*
+import platform.posix.*
+import libcurl.*
+
+internal class CurlRequest(val newRequests: Set<CurlRequestData>, override val listenerKey: ListenerKey) : WorkerRequest
+
+internal class CurlResponse(val completeResponses: Set<CurlResponseData>, override val listenerKey: ListenerKey) :
+    WorkerResponse
+
+// Only set in curl thread
+@ThreadLocal
+private var curlState: CurlState? = null
```

Comment:
Could we avoid using shared state?
What could happen if we run multiple curl engines in a single thread?